File: ts-src/commands/registry.ts
import { findClosest } from "../tools/string-distance.ts";
import type { SlashCommand } from "./types.ts";

export class CommandRegistry {
	private commands = new Map<string, SlashCommand>();

	register(command: SlashCommand): void {
		this.commands.set(command.name, command);
	}

	get(name: string): SlashCommand | undefined {
		return this.commands.get(name);
	}

	all(): SlashCommand[] {
		return [...this.commands.values()];
	}

	suggest(name: string): string | undefined {
		const candidates = [...this.commands.keys()];
		return findClosest(name, candidates) ?? undefined;
	}
}

File: src/commands/registry.rs
//! Command registry with lookup + suggestion.

use std::collections::HashMap;

use crate::tools::string_distance::find_closest;

use super::types::SlashCommand;

#[derive(Default)]
pub struct CommandRegistry {
    commands: HashMap<String, SlashCommand>,
}

impl CommandRegistry {
    pub fn new() -> Self {
        Self::default()
    }
    pub fn register(&mut self, command: SlashCommand) {
        self.commands.insert(command.name.clone(), command);
    }
    pub fn get(&self, name: &str) -> Option<&SlashCommand> {
        self.commands.get(name)
    }
    pub fn all(&self) -> Vec<&SlashCommand> {
        self.commands.values().collect()
    }
    pub fn suggest(&self, name: &str) -> Option<String> {
        let candidates: Vec<String> = self.commands.keys().cloned().collect();
        find_closest(name, &candidates, 3).map(String::from)
    }
}
